### PR TITLE
Make the balance roomier

### PR DESCRIPTION
**Pull request by selalipop**
This PR:

- Makes `Purchase` objects able to represent purchases and deposits
- Migrates the DB to reflect this new change
- Uses Room as a single source of truth for balance

### CODE REVIEW

File: app/src/main/java/com/example/allowancetracker/data/Converters.kt [hunk 1]
```diff
@@ -13,4 +13,11 @@ class Converters {
     fun dateToTimestamp(date: Date?): Long? {
         return date?.time
     }
+
+    @TypeConverter
+    fun toPurchaseType(purchaseType: PurchaseType): Int = purchaseType.ordinal
```
Comment (SpudLaw) [inline app/src/main/java/com/example/allowancetracker/data/Converters.kt]: what is an ordinal? :]

File: app/src/main/java/com/example/allowancetracker/ui/main/MainFragment.kt [hunk 2]
```diff
@@ -68,7 +68,7 @@ class MainFragment : Fragment() {
                     if (costEditText.text?.isNotBlank() == true && descriptionEditText.text?.isNotBlank() == true) {
                         val cost: Double = costEditText.text?.toString()?.toDouble() ?: 0.0
                         val description: String = descriptionEditText.text.toString()
-                        val purchase = Purchase(PurchaseType.Purchase, cost, null, description)
+                        val purchase = Purchase(PurchaseType.Purchase, cost, Date(), description)
```
Comment (SpudLaw) [inline app/src/main/java/com/example/allowancetracker/ui/main/MainFragment.kt]: 💯

File: app/src/main/java/com/example/allowancetracker/data/Purchase.kt [hunk 3]
```diff
@@ -12,7 +12,7 @@ import java.util.*
 data class Purchase(
     @ColumnInfo(name = "type") val type: PurchaseType,
     @ColumnInfo(name = "cost") val cost: Double,
-    @ColumnInfo(name = "date") val date: Date?,
+    @ColumnInfo(name = "date") val date: Date? = Date(),
```
Comment (SpudLaw) [inline app/src/main/java/com/example/allowancetracker/data/Purchase.kt]: 👏🏼

File: app/src/main/java/com/example/allowancetracker/data/PurchaseDatabase.kt [hunk 4]
```diff
@@ -10,7 +10,7 @@ import androidx.sqlite.db.SupportSQLiteDatabase
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.launch
 
-@Database(entities = [Purchase::class], version = 2)
+@Database(entities = [Purchase::class], version = 3)
```
Comment (SpudLaw) [inline app/src/main/java/com/example/allowancetracker/data/PurchaseDatabase.kt]: 😭 thank you for doing this for meeeeee!!!

File: app/src/main/java/com/example/allowancetracker/data/Converters.kt (hunk 1 above)
Comment (selalipop) [reply app/src/main/java/com/example/allowancetracker/data/Converters.kt]: Ordinal is a not-so-clear 👀 built in name for the `Int` that the enum value converts to

Like:
`enum Color{
   Red,     // Ordinal 0
   Green, // Ordinal 1
   Blue    //Ordinal 2
}

`

File: app/src/main/java/com/example/allowancetracker/data/PurchaseDatabase.kt (hunk 4 above)
Comment (selalipop) [reply app/src/main/java/com/example/allowancetracker/data/PurchaseDatabase.kt]: no problemo 😁

File: app/src/main/java/com/example/allowancetracker/data/Converters.kt (hunk 1 above)
Comment (SpudLaw) [reply app/src/main/java/com/example/allowancetracker/data/Converters.kt]: oooo okay thank you! :D

File: app/src/main/java/com/example/allowancetracker/ui/main/MainViewModel.kt [hunk 5]
```diff
@@ -1,70 +1,43 @@
...
+    val balance = purchases.map {
+        //Each time we get a new list of purchases
+        //Starting with a 0.0 balance, loop through the list
+        it.fold(0.0) { total, purchase ->
+            when (purchase.type) {
+                //If a purchase is a deposit, we add it to the total
+                PurchaseType.InitialDeposit,
+                PurchaseType.Deposit -> total + purchase.cost
+                //If a purchase is a purchase, we remove it from the total
+                PurchaseType.Purchase -> total - purchase.cost
```
Comment (SpudLaw) [inline app/src/main/java/com/example/allowancetracker/ui/main/MainViewModel.kt]: this is so great! so clean and small i love it thank you :3

**Review by SpudLaw — APPROVED** [review]

**Pull request merged**